File: src/main/java/com/company/erp/features/users/service/UserService.java
package com.company.erp.features.users.service;

import com.company.erp.core.database.PageQuery;
import com.company.erp.core.exceptions.ConflictException;
import com.company.erp.core.exceptions.NotFoundException;
import com.company.erp.features.users.dto.AssignRolesRequest;
import com.company.erp.features.users.dto.CreateUserRequest;
import com.company.erp.features.users.dto.UpdateUserRequest;
import com.company.erp.features.users.entity.Role;
import com.company.erp.features.users.entity.User;
import com.company.erp.features.users.repository.RoleRepository;
import com.company.erp.features.users.repository.UserRepository;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Sort;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.HashSet;
import java.util.List;
import java.util.Set;
import java.util.stream.Collectors;

@Service
@Transactional
public class UserService {

    private static final Set<String> ALLOWED_SORT = Set.of("email", "fullName", "createdAt");

    private final UserRepository users;
    private final RoleRepository roles;
    private final PasswordEncoder passwords;

    public UserService(UserRepository users, RoleRepository roles, PasswordEncoder passwords) {
        this.users = users;
        this.roles = roles;
        this.passwords = passwords;
    }

    @Transactional(readOnly = true)
    public Page<User> list(PageQuery query) {
        // Search filtering wires in once the products module introduces JPA
        // Specifications as the canonical pattern. For now: paginate only.
        return users.findAllBy(query.toPageable(ALLOWED_SORT, Sort.by("email")));
    }

    @Transactional(readOnly = true)
    public User getById(Long id) {
        return users.findWithRolesById(id).orElseThrow(() -> new NotFoundException("User not found"));
    }

    @Transactional(readOnly = true)
    public User getByEmail(String email) {
        return users.findByEmail(email.toLowerCase()).orElseThrow(() -> new NotFoundException("User not found"));
    }

    public User create(CreateUserRequest req) {
        String email = req.email().toLowerCase();
        if (users.existsByEmail(email)) throw new ConflictException("Email already in use");

        User user = new User();
        user.setEmail(email);
        user.setPasswordHash(passwords.encode(req.password()));
        user.setFullName(req.fullName());
        user.setPhone(req.phone());
        user.setRoles(new HashSet<>(resolveRoles(req.roles())));
        return users.save(user);
    }

    public User update(Long id, UpdateUserRequest req) {
        User user = getById(id);
        if (req.fullName()  != null) user.setFullName(req.fullName());
        if (req.phone()     != null) user.setPhone(req.phone());
        if (req.avatarUrl() != null) user.setAvatarUrl(req.avatarUrl());
        if (req.enabled()   != null) user.setEnabled(req.enabled());
        if (req.roles()     != null) user.setRoles(new HashSet<>(resolveRoles(req.roles())));
        return user;
    }

    public void delete(Long id) {
        users.delete(getById(id));
    }

    public List<User> assignRoles(AssignRolesRequest req) {
        List<User> targets = users.findAllWithRolesByIdIn(req.userIds());
        if (targets.size() != req.userIds().size()) {
            Set<Long> found = targets.stream().map(User::getId).collect(Collectors.toSet());
            Set<Long> missing = req.userIds().stream()
                    .filter(id -> !found.contains(id))
                    .collect(Collectors.toSet());
            throw new NotFoundException("User(s) not found: " + missing);
        }

        Set<Role> resolved = new HashSet<>(resolveRoles(req.roles()));
        for (User user : targets) {
            switch (req.mode()) {
                case ADD     -> user.getRoles().addAll(resolved);
                case REMOVE  -> user.getRoles().removeAll(resolved);
                case REPLACE -> user.setRoles(new HashSet<>(resolved));
            }
        }
        return targets;
    }

    private Set<Role> resolveRoles(Set<String> codes) {
        if (codes.isEmpty()) return Set.of();
        return codes.stream()
                .map(code -> roles.findByCode(code)
                        .orElseThrow(() -> new NotFoundException("Role not found: " + code)))
                .collect(Collectors.toSet());
    }
}

File: app/Features/Users/Services/UserService.php
<?php

namespace App\Features\Users\Services;

use App\Features\Users\Models\Role;
use App\Features\Users\Models\User;
use App\Support\Exceptions\ConflictException;
use App\Support\Exceptions\NotFoundException;
use App\Support\Pagination\PageQuery;
use Illuminate\Contracts\Pagination\LengthAwarePaginator;
use Illuminate\Support\Facades\DB;
use Illuminate\Support\Facades\Hash;

class UserService
{
    /** @var array<string, string> apiField => dbColumn */
    private const ALLOWED_SORT = [
        'email' => 'email',
        'fullName' => 'full_name',
        'createdAt' => 'created_at',
    ];

    public function list(PageQuery $query): LengthAwarePaginator
    {
        // Search filtering wires in with the products module's filter pattern;
        // for now this paginates only, matching the original UserService.
        $builder = User::query()->with('roles.permissions');
        $query->applySort($builder, self::ALLOWED_SORT, ['email', 'asc']);

        return $builder->paginate($query->pageSize, ['*'], 'page', $query->page);
    }

    public function getById(int $id): User
    {
        $user = User::query()->with('roles.permissions')->find($id);
        if (! $user) {
            throw new NotFoundException('User not found');
        }
        return $user;
    }

    public function getByEmail(string $email): User
    {
        $user = User::query()->with('roles.permissions')
            ->where('email', strtolower($email))->first();
        if (! $user) {
            throw new NotFoundException('User not found');
        }
        return $user;
    }

    /**
     * @param array<string, mixed> $data  email,password,fullName,phone,roles
     */
    public function create(array $data): User
    {
        $email = strtolower($data['email']);
        if (User::query()->where('email', $email)->exists()) {
            throw new ConflictException('Email already in use');
        }

        return DB::transaction(function () use ($email, $data) {
            $user = new User();
            $user->email = $email;
            $user->password_hash = Hash::make($data['password']);
            $user->full_name = $data['fullName'];
            $user->phone = $data['phone'] ?? null;
            $user->save();

            $roles = $this->resolveRoles($data['roles'] ?? []);
            $user->roles()->sync($roles->pluck('id')->all());

            return $user->load('roles.permissions');
        });
    }

    /**
     * @param array<string, mixed> $data  partial: fullName,phone,avatarUrl,enabled,roles
     */
    public function update(int $id, array $data): User
    {
        return DB::transaction(function () use ($id, $data) {
            $user = $this->getById($id);

            if (array_key_exists('fullName', $data) && $data['fullName'] !== null) {
                $user->full_name = $data['fullName'];
            }
            if (array_key_exists('phone', $data) && $data['phone'] !== null) {
                $user->phone = $data['phone'];
            }
            if (array_key_exists('avatarUrl', $data) && $data['avatarUrl'] !== null) {
                $user->avatar_url = $data['avatarUrl'];
            }
            if (array_key_exists('enabled', $data) && $data['enabled'] !== null) {
                $user->enabled = (bool) $data['enabled'];
            }
            $user->save();

            // null means "leave roles untouched"; present (even empty) replaces.
            if (array_key_exists('roles', $data) && $data['roles'] !== null) {
                $roles = $this->resolveRoles($data['roles']);
                $user->roles()->sync($roles->pluck('id')->all());
            }

            return $user->load('roles.permissions');
        });
    }

    public function delete(int $id): void
    {
        $user = $this->getById($id);
        $user->delete();
    }

    /**
     * @param array<int, int> $userIds
     * @param array<int, string> $roleCodes
     * @return array<int, User>
     */
    public function assignRoles(array $userIds, array $roleCodes, string $mode): array
    {
        return DB::transaction(function () use ($userIds, $roleCodes, $mode) {
            $targets = User::query()->with('roles.permissions')
                ->whereIn('id', $userIds)->get();

            if ($targets->count() !== count($userIds)) {
                $found = $targets->pluck('id')->map('intval')->all();
                $missing = array_values(array_diff($userIds, $found));
                throw new NotFoundException('User(s) not found: ['.implode(', ', $missing).']');
            }

            $resolved = $this->resolveRoles($roleCodes);
            $roleIds = $resolved->pluck('id')->all();

            foreach ($targets as $user) {
                switch ($mode) {
                    case 'ADD':
                        $user->roles()->syncWithoutDetaching($roleIds);
                        break;
                    case 'REMOVE':
                        $user->roles()->detach($roleIds);
                        break;
                    case 'REPLACE':
                        $user->roles()->sync($roleIds);
                        break;
                }
                $user->load('roles.permissions');
            }

            return $targets->all();
        });
    }

    /**
     * @param array<int, string> $codes
     * @return \Illuminate\Support\Collection<int, Role>
     */
    private function resolveRoles(array $codes)
    {
        if (empty($codes)) {
            return collect();
        }
        return collect($codes)->map(function ($code) {
            $role = Role::query()->where('code', $code)->first();
            if (! $role) {
                throw new NotFoundException('Role not found: '.$code);
            }
            return $role;
        });
    }
}
